packages/melody-hooks/src/component.js

```diff
@@ -0,0 +1,353 @@
+     * This might cause the component to request an update.
+     * @param {*} props new properties
+     */
+    apply(props) {
+        if (shallowEquals(props, this.props)) {
+            return;
+        }
+        this.props = props || null;
+        this.isPropsDirty = true;
+        this.enqueueComponent();
+    },
+
+    /**
+     * Executed after a component has been mounted or updated.
+     * After this method has been triggered, the component is
+     * considered stable and accessing the DOM should be safe.
+     * The children of this Component might not have been rendered.
+     */
+    notify() {
+        if (this.isMounted) {
+            if (options.afterUpdate) afterUpdate(this);
+        } else {
+            this.isMounted = true;
+            if (afterMount) afterMount(this);
+        }
+
+        const hooks = this.hooks;
+
+        // Mark this component as in running effects
+        // This is needed in order to not immediately enqueue
+        // the component again when `setState` is called while
+        // we are running the effects
+        this.isRunningEffects = true;
+
+        // Run through hooks and call effects
+        for (let i = 0, l = hooks.length; i < l; i++) {
+            const hook = hooks[i];
+            const type = hook[0];
+
+            // We are only interested in effect hooks
+            if (type !== HOOK_TYPE_USE_EFFECT) {
+                continue;
+            }
+
+            const dirty = hook[3];
+
+            // If this effect hook is not marked as
+            // `dirty` we can skip the invocation.
+            // A hook is considered `dirty` when the
+            // input values have changed
+            if (!dirty) continue;
+
+            const callback = hook[1];
+            const unsubscribe = hook[4];
+
+            // If the hook returned an `unsubscribe`
+            // function the first time it was called,
+            // we need to invoke it first
+            if (unsubscribe) {
+                unsubscribe();
+            }
+
+            // Call the effect and store a potential
+            // `unsubscribe` function for later use
+            const unsubscribeNext = callback() || null;
+            hook[4] = unsubscribeNext;
+        }
+
+        // Reset `isRunningEffects`
+        this.isRunningEffects = false;
+
+        // If `setState` was called during the effects,
+        // we enqueue the component for rendering.
+        if (this.isStateDirty) {
+            this.enqueueComponent();
+        }
+    },
+
+    /**
+     * Enqeues a state change for a hook slot at index `hookIndex`
+     * @param {*} hookIndex Index of the `useState` hook
+     * @param {*} valueNext Updated value for this state hook
+     * @return void
+     */
+    setState(hookIndex, valueNext) {
+        // Store the new state in the queue
+        const finalValue =
+            typeof valueNext === 'function'
+                ? valueNext(this.state[hookIndex])
```

Comment:
What happens if we call `setState` twice or more? I think we are providing the wrong value to the function `valueNext`.
E.g.
```js
function myComponentFn(props) {
  const [val, setter] = useState(0);
  setter((prev) => prev + 1);  // State is updated to `1`
  setter((prev) => prev + 1);  // State is updated again to `1` but I was expecting `2`
  return props;
}
```
I'm not sure how React handles these situations. And I think it's not so uncommon either: the user may perform multiple `setState` across different `effects`.
What's your opinion?

Reply:
If you call it like that inside the component function we throw the "too many re-renders" error. If you use it in an effect `val` would be `2`

Reply:
It should actually be
```js
const finalValue =
  typeof valueNext === 'function'
    ? valueNext(this.stateQueue[hookIndex] || this.state[hookIndex])
    : valueNext
```

Then `val` would be `2`

Reply:
Yes, that would actually do the trick! But I would be stricter with the check since the user can potentially set `0` or `''` to the state.

Reply:
What do you mean by stricter?

Reply:
```js
const prevState = hookIndex in this.stateQueue
  ? this.stateQueue[hookIndex]
  : this.state[hookIndex];

valueNext(prevState);
```

Reply:
ah ok, true!

Reply:
fixed

Reply:
What about `false` as a value? If that was within `stateQueue`, then we'd also take the old value from `state`.
I guess you'll need to check if a new value has been queued.

Reply:
That's correct, it has been fixed. 👍

Reply:
wait this wouldn't work. you are right @pago 
```js
setState(state => undefined);
setState(state => state + 1);
```

Reply:
Should've reloaded the page before posting. :)

Reply:
Generally `undefined` is not considered as a valid state. So it's not critical imho.